### ISSUE TITLE
buildGraalvmNativeImage causing overrideAttrs to be ignored

### ISSUE THREAD

**Issue by jakubgs** [NixOS/nixpkgs#216787]
### Describe the bug
While trying to upgrade `zprint` Clojure formatting tool using `overrideAttrs` I discovered that it's not effective when changing `src`. The `buildGraalvmNativeImage` builder from [`pkgs/build-support/build-graalvm-native-image/default.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/build-graalvm-native-image/default.nix) simply ignores it and builds the old version provided to it. Another wrapper which has this behavior is `buildGoModule` from [`pkgs/build-support/go/module.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix), except in that case build fails due to wrong versions in `vendor` folder.

It seems like those builders/wrappers around `mkDerivation` make it impossible to override things like `src` via `overrideAttrs` like for projects that simply use `mkDerivation` without using special syntax specific for those wrappers. Not sure if this is intentional, but it certainly is a surprise when you expect `overrideAttrs` to work.

### Steps To Reproduce

1. Attempt to override `version` and `src` of `gomobile` package:
    ```nix
    zprint = super.zprint.overrideAttrs (old: rec {
      inherit (old) pname;
      version = "1.2.5";
      src = self.fetchurl {
        url = "https://github.com/kkinnear/${pname}/releases/download/${version}/${pname}-filter-${version}";
        sha256 = "sha256-PWdR5jqyzvTk9HoxqDldwtZNik34dmebBtZZ5vtva4A=";
      };
    });
    ```
3. Build the overridden derivation:
    ```
     > nix-build
    /nix/store/8qb7p2k0d53llarndkdjz5n25lly8777-zprint-1.2.5
    ```
4. Check binary version:
    ```
     > /nix/store/8qb7p2k0d53llarndkdjz5n25lly8777-zprint-1.2.5/bin/zprint --version
    zprint-1.2.4
    ```
4. Check the inputs for the derivation:
    ```
     > nix show-derivation /nix/store/8qb7p2k0d53llarndkdjz5n25lly8777-zprint-1.2.5 | grep nativeImageBuildArgs
          "buildPhase": "runHook preBuild\n\nnative-image ${nativeImageBuildArgs[@]}\n\nrunHook postBuild\n",
          "extraNativeImageBuildArgs": "--no-server -H:EnableURLProtocols=https,http -H:+ReportExceptionStackTraces --report-unsupported-elements-at-runtime --initialize-at-build-time --no-fallback",
          "nativeImageBuildArgs": "-jar /nix/store/qnbb6926l2v7b11hwbhksxchxmjr8c23-zprint-filter-1.2.4 -H:CLibraryPath=/nix/store/6jg566zdnvpd65i0llrblki0qiyqcsnx-graalvm11-ce-22.2.0-lib/lib  -H:Name=zprint --verbose --no-server -H:EnableURLProtocols=https,http -H:+ReportExceptionStackTraces --report-unsupported-elements-at-runtime --initialize-at-build-time --no-fallback -J-Xmx6g",
    ```
6. See that the JAR used is still that of `1.2.4` version.
    ```
    -jar /nix/store/qnbb6926l2v7b11hwbhksxchxmjr8c23-zprint-filter-1.2.4
    ```

### Expected behavior
I'd expect the `overrideAttrs` to work for all packages, but it doesn't.

### Additional context
The workaround in those cases is to use `override` and apply the change to the builder directly, for example:
```nix
zprint = super.zprint.override rec {
  buildGraalvmNativeImage = args: super.buildGraalvmNativeImage ( args // rec {
    inherit (args) pname;
    version = "1.2.5";
    src = self.fetchurl {
      url = "https://github.com/kkinnear/${pname}/releases/download/${version}/${pname}-filter-${version}";
      sha256 = "sha256-PWdR5jqyzvTk9HoxqDldwtZNik34dmebBtZZ5vtva4A=";
    };
  });
};
```
Which builds correctly, but is an unintuitive and surprising way of doing this.

### Notify maintainers
@thiagokokada

**Comment by thiagokokada** [comment]
I  actually stumbled this while developing #216228. Decided not to fix there but it should be a fairly easy fix.

**Comment by Uthar** [comment]
I noticed this when trying to override Go packages. Python modules have the `overridePythonAttrs` for this purpose. Maybe other language frameworks should do that too?

**Comment by thiagokokada** [comment]
> I noticed this when trying to override Go packages. Python modules have the `overridePythonAttrs` for this purpose. Maybe other language frameworks should do that too?

This bug was fixed in #216855. I think you may open another bug for Go packages.

Closing this ticket.
